Restructure expression syntax guide and add missing content (#105)
wso2/docs-choreo-dev: en/docs/references/ballerina-expression-syntax.md

# Ballerina Expression Syntax

> Ballerina expressions comprise a powerful expression syntax based on the [Ballerina language](https://ballerina.io/) for function invocations, declaring a variable using literals and tuples, typecasting, etc. This document presents you with the common expressions that you can use while developing in the Choreo console.

Author: yeah, that makes sense. Changed.

## Variables

> A variable statement holds a value of a specific data type (`string`, `integer`, etc.).  You can declare a variable and use it later in the logical process of your service or the integration.

Author: changed.

> ### var

Author: What do you mean overusing `var`? This is a section under variables.

> ### var

Author: ![Screenshot from 2021-06-04 12-47-38](https://user-images.githubusercontent.com/7173347/120761681-14543900-c533-11eb-90fb-613d81393910.png)

> ### var

Reviewer: Sorry, I meant we don't encourage overusing `var` from the language perspective. Was wondering if we should first explain usage with the types and make `var` the last one in the list.

> ### var

Author: This was placed first, as it is the first and the default type selected in the Choreo Console when adding a variable. Will check with the team and do the needful.

> ### var

Author: Changed.

> Ballerina allows you to declare variables using the `var` keyword instead of specifying a type. The expression on the right-hand side decides the type of the declared variable.

Author: We can't use this exact sentence as it is passive voice. We write our documents in active voice. Hence, re phrased this paragraph to "Ballerina allows you to declare variables using the `var` keyword. Here, Ballerina infers the type of the declared variable based on the right-hand side expression."

###### Examples:

> To initialize a `var` in Choreo, you can use one of the following;

Author: This is for `var`

> To initialize a `var` in Choreo, you can use one of the following;

Reviewer: Then shall we say

```
To initialize a variable declared with `var` in Choreo, you can use one of the following;
```

> To initialize a `var` in Choreo, you can use one of the following;

Author: yes, Changed.

```ballerina
    5
    
   "Hello!"
```

### int

> An integer is a positive number, a negative number, or zero that does not include a fraction. A signed integer is an integer with the  `+` or `-` character in the front, indicating the sign. You can also define a positive integer or zero using only numerical characters. 

Author: They seem to be more spec oriented. Used the definitions in the https://ballerina.io/learn/by-example page, which is  more user-friendly.

###### Examples:

> To initialize an `int` in Choreo, you can use one of the following;

Author: agreed. changed.

  ```ballerina
  1234
  
  -10
  
  +2020
  ```

> ### float

Author: In the choreo console, we cannot add a variable type decimal. Do you think we should still mention it in the description? Would it confuse the user?
Currently we define float as "The `float` type is IEEE 64-bit binary floating point (same as double in Java) and supports the same arithmetic operators as `int`. You have the option to use `+` or `-` characters in the front to indicate the sign."

> ### float

Reviewer: Oh okay. A value accessed from a JSON also may be a decimal. Wonder if that could be an issue,

A float is a number with a decimal point. You have the option to use `+` or `-` characters in the front to indicate the sign.

###### Examples:

To initialize a `float` in Choreo, you can use one of the following;
  ```ballerina
  0.3353
  
  -8593.992
  
  +950.930
  ```

### boolean

The boolean data type has one of the two possible values, `true` or `false`.

### string

A string is a sequence of characters. It is required to use double quotes (`"`)  to mark the boundary between the starting and ending characters.

###### Examples:

> To initialize a `float` in Choreo, you can use one of the following;

Author: changed.

  ```ballerina
  "Ballerina"
  
  "Ballerina is a programming language intended for network-distributed applications."
  ```

## Value/type operations

### String operations

- Concatenation

    You can use the `+` operator to do string concatenations. The string value can come from a string literal or a variable.

    ###### Example:

    To concatenate the string `Hi` and a string variable `name`, you can use the following expression:

    ```ballerina
     "Hi " + name;
    ```

> ### Arithmetic operators

Author: Done.

> Arithmetic operators perform addition, subtraction, multiplication, division, and modules on numerical values. You can use literals or variables of type `int` or `float` as the numerical values in Ballerina expressions.

Author: Yes, changed to modulo.
We don't have a variable type decimal. Hence, didn't use it.

- Addition

    You can use the `+` operator for number addition. 

    ###### Example:

    To add an integer literal (`20`) and an integer variable `number1` and store it in an integer variable `sum`, you can use the following expression:

    ```ballerina

>     number1 + 20;

Author: Modified.

    ###### Example:

    To subtract two integers, integer variable `yearOfBirth` from integer literal `2022` and print the result, you can use the following expression:

    ```ballerina
    2022 - yearOfBirth;
    ```
 
- Division

    You can use the `/` operator for number division. 

    ###### Example:

    To divide an integer variable `totalNumberOfStudents` from integer literal `5` and store it in an integer variable `numberOfStudentsPerClass`, you can use the following expression:

    ```ballerina
      totalNumberOfStudents / 5; 
    ```
 
- Multiplication

    You can use the `*` operator for number multiplication. 

    ###### Example:

    To multiply two float constants `60` by `5` and store it in a float variable `numberOfMinsIn5Hours`, you can use the following expression:

    ```ballerina
      60 * 5;
    ```

- Modulo 

    You can use the  `%` operator to get the remainder of a division. 

    ###### Example:

    To get the remainder of the division between an integer variable `n` and an integer literal `3`, you can use the following expression:
  
    ```ballerina
    (n % 3) > 2
    ```

### Value/type operators

- Equality 

>     You can use the equality operators to check the equality of values of any of the [Ballerina language's](https://ballerina.io/) basic types (`string`, `int`, `float`, `decimal`, `boolean`, and `nil`). These values can come from a literal or a variable. **Note** that the resulting value from these operators is always of `boolean` type.   

Author: Changed.

    
    Ballerina provides the following equality operators: 
    

>     - `==`: For deep value equality of anydata|error typed values.

Author: changed.

    - `===`: For reference equality. 
    - `!= `and `!==`: For negations.

    ###### Examples:

    To compare an integer literal `10` with an integer variable `firstNumber`, you can use the following expression:
    
    ```ballerina
    10 == firstNumber
    ```

    To compare a string variable `username`  with  a string literal `admin`, use the following expression:  

    ```ballerina
    username != "admin" 
    ```

- Numeric value comparison

>     The following operators are available in the expression editor for numeric value comparisons:

Author: Generalized it. What are the ordered types in Ballerina?  If so we can explicitly mention it.

    
    - `<`: Less than
    - `>`: Greater than
    - `<=`: Less than or equal to
    - `>=`: Greater than or equal to

    ###### Examples:

    To compare the value of the integer variable `firstNumber` to the integer literal `10`, use the following expressions:

    ```ballerina
    firstNumber > 10
  
    firstNumber <= 10
    ```

## HTTP 

> All the HTTP-related service applications created in Choreo have a variable called `request` (of type `http:Request`) which the developers can use to query information related to the HTTP request. Following is a list of such use cases.

Author: done

### Read a header value in the request

> A typical HTTP request contains many headers.  These include standard and custom headers. You can use the `request` variable's `getHeader` function to read a specific header value in any expression-enabled input field.

Author: Done.

###### Examples:

To read the value of an HTTP header `Access-Control-Allow-Origin`, use the following expression:

  ```ballerina
  check request.getHeader("Access-Control-Allow-Origin")
  
  ```
  
To read the value of an HTTP header by passing a variable that contains the header name `headerName`, use the following expression:

  ```ballerina
  check request.getHeader(headerName)
  ```

### Check whether a header is present in the request

> Applications sometimes need to check the availability of a header before proceeding to something like reading the header value. You can use the  `hasHeader`  function in the `request` object for this purpose.

Author: Changed.

###### Examples:

To check if the HTTP header `content-length` exists given the header name, use the following expression:

  ```ballerina
  request.hasHeader("content-length")
  ```
  
To check if the HTTP header `content-length` exists by passing the header name in the string variable `contentLength`, use the following expression: 

  ```ballerina
  request.hasHeader(contentLength)
  ```
### Read the JSON Payload of the request

> JSON is a common content type used for HTTP communication. You can use the `request` variable’s `getJsonPayload` function to read the JSON payload sent with the HTTP request. Note that the `getJsonPayload` function’s return type is `json|ClientError`.

Author: fixed.

###### Example:

To read the JSON Payload of the HTTP request, use the following expression:

  ```ballerina
  request.getJsonPayload()
  ```

### Read the Text Payload of the request

> Sometimes, content is sent as a string (plain text) with the HTTP request. You can use the `request` variable’s `getTextPayload` function to read the text payload sent with the HTTP request. Note that the `getTextPayload` function’s return type is `string|ClientError`.

Author: fixed.

###### Example:

To read the text payload of the HTTP request, use the following expression:

  ```ballerina
  check request.getTextPayload()
  ```

### Read the value of a query parameter 

> Query parameters are used in an HTTP request to send additional inputs to consider when processing the request. You can use the `request` variable's `getQueryParamValue` function to read the passed query parameter.

Author: fixed.

###### Examples:

To read the value of the query parameter `category`, use the following expression:

  ```ballerina
  request.getQueryParamValue("category")
  ```
  
To read the value of the query parameter passing the query parameter name as a variable, use the following expression:

  ```ballerina

>   request.getQueryParamValue(queryParamName)

Author: It would add value wouldn't it? No one would read this from end to end, and would only refer to a section. So, I think it would be helpful.

  ```

### Read the cookies available in the request

> HTTP cookies keep stateful information against a client. You can use the `request` variable's `getCookies` function to read cookies present in the request as an array.

Author: fixed.

###### Example:

To read all the cookies available in the request, use the following expression:

  ```ballerina
  request.getCookies()
  ```

## JSON

JSON is a textual format for representing a collection of values: a simple value, an array of values, or an object. Ballerina has a single type named `json` that can represent any JSON value. Thus, `json` is a built-in union type in Ballerina that can contain values of the types `nil` (as the null value), `boolean`,` int`,` float`, `decimal`, `string`, `json[]`, and `map<json>`.

### Simple values in JSON 

> Ballerina has a single type named `json` that can represent any JSON value. You can use `json` which is a built-in union type in Ballerina, to store values of the types `nil` (as the null value), `boolean`, `int`, `float`, `decimal`, `string`, `json[]`, and `map<json>`. 

Author: Fixed.

###### Examples:

> To initialize a `json` type variable, you can use the following:

Author: fixed.

```ballerina
    "Alex"

    5.36
    
    false
    
    [1, false, null, "foo", {first: "Alex", last: "John"}]
    
    {name: "Alex", age: null, marks: {math: 90, language: 95}}
    
     null
```

### JSON Objects

> Ballerina represents JSON objects as maps. You can define a JSON object to hold values of any `json` type (`string`, `number`, `boolean`, `object`, `array`, or `null`). A JSON object can be declared in two ways:

Reviewer: ```suggestion
Ballerina represents JSON objects as maps. You can define a JSON object to hold any value that belongs to the `json` type (string, number, boolean, object, array, or null). A JSON object can be declared in two ways:
```

>     -  Using the variable type `json`.

Reviewer: `json` is not just used for JSON objects. The following is also valid.

```ballerina
json j = 1;
```

Wonder if this could cause confusion?

>     -  Using the variable type `other` and `map<json>` as the `other type`.

Reviewer: Emphasizing `other` and `other type` with backticks kinda gives the impression that they are types IMO. Can we use quotes maybe?
```suggestion
    -  Selecting "other" and specifying `map<json>` as the "other type" for the type of the variable.
```

###### Examples:

To declare a JSON object, use the following:

```ballerina

>      {name: "apple", color: "red", price: "20.0"}    

Reviewer: To show that it can have values of different types should we do

```suggestion
     {name: "apple", color: "red", price: 20.0}    
```

```

To create an empty JSON object, use the following:

```ballerina
    {}
```

> To cast a `json` type variable `fruitDetail` to a `map<json>` type variable, use the following:

Reviewer: ```suggestion
To cast a variable name `fruitDetail` of type `json` to the `map<json>` type, use the following:
```

```ballerina

<map<json>>fruitDetail
 
```

To access or change a field of the JSON object `fruitDetailObject`, use one of the following:

!!! note
    If you define the JSON object using a `json` type variable, you need to cast it to a `map<json>` first.
    
```ballerina
   fruitDetailObject["color"] = "orange"; 
   

>    fruitDetailObject["name"];

Reviewer: To access you can also use `fruitDetailObject.name` even when fruitDetailObject is not map<json>.

```

To create a nested JSON object, use the following:

```ballerina
    map<json> family = {
           fname: "Alex",
           lname: "Stallone",
           address: {
               line: "20 Palm Grove",
               city: "Colombo 03",
               country: "Sri Lanka"
           }
       };
```

### JSON Arrays

> JSON array values must be of any of the json types: `string`, `number`, `boolean`, `object`, `array`, or `null`. You can define a JSON array in one of the following two ways:

Reviewer: ```suggestion
Values in a JSON array must be of one of the JSON types: string, number, boolean, object, array, or null. You can define a JSON array in one of the following two ways:
```

>     - Using a [`json` type variable](#simple-values-in-JSON).  

Reviewer: ```suggestion
    - Using a [variable of type `json`](#simple-values-in-JSON).  
```

>     - Using the variable type `other` and `json[]` as the `other type`.

Reviewer: Same comment as the previous one re: emphasis.

JSON array elements can be accessed by index.

###### Examples:

To declare a JSON array, use the following:

```ballerina
 [1, false, null, "foo", {first: "Alex", last: "John"}]
 
```

> To cast the `json` type variable `personRecord` to a `json[]` type variable, use the following:

Reviewer: ```suggestion
To cast a variable named `personRecord` of the `json` type  to the `json[]` type, use the following:
```

```ballerina
<json[]>personRecord;
```

> To access elements the 4th element in the JSON array `personRecordArray`, use one of the following:

Reviewer: ```suggestion
To access the 4th element in the JSON array `personRecordArray`, use the following:
```

!!! note
    If you define the JSON array using a `json` type variable, you need to cast it to a `json[]` type first.
    
```ballerina
personRecordArray[4]; 
```

To add a JSON array in an object literal, use the following:

```ballerina
    {
        fname: "Alex",
        lname: "Stallone",
        family: [
            {fname: "James", lname: "Stallone"},
            {fname: "Blake", lname: "Stallone"},
            {fname: "Kyle", lname: "Stallone"},
            {fname: "Taylor", lname: "Stallone"}
        ]
    }
```

To get the length of a JSON array, use the following:

```ballerina
personRecordArray.length(); 
```

### JSON access

> You can use the field access(.) and optional field access (?.) to access the `json` type in Ballerina. In this case, the type checking happens at runtime. The return type of this operation is a union of `json` and `error`. 

Reviewer: ```suggestion
You can use field access(.) and optional field access (?.) to access values from a variable of type `json` in Ballerina. In this case, the type checking happens at runtime. The return type of this operation is the union of the member type and `error`. 
```

Reviewer: Where `x` is `map<map<json>> x = {}` the type of `x.a` is `map<json>|error`.

    
!!! info

>     The examples in this section will assume the following json is already defined.

Reviewer: ```suggestion
    The examples in this section will assume the following variable of type `json` is already defined.
```

    
    ```ballerina
         json person = {
             fname: "Alex",
             lname: "Jane",
             address: {
                 line: "20 Palm Grove",
                 city: "Colombo 03",
                 country: "Sri Lanka"
             }
         }
    ```    

- Field access(.) 

>     You can use the field access operator to access record fields of the `json` type. If the key is invalid or the field access occurs on `null`, it returns `error`.

Reviewer: ```suggestion
    You can use field access to access values from a variable of type `json` that may hold a JSON object. If the value is not a JSON object or if the key is not present, field access returns an error at runtime.
```

Author: Fixed.

   
    ###### Examples:

    To access a field, use the following:    

    ```ballerina
       check person.fname;
    ```
    To access a field in a nested level, use the following:
 
     ```ballerina
        check person.address.city;
     ```

- Optional field access (?.) 

>     You can use the optional field access operator to access record fields, including any optional fields. If the key is invalid or the field access occurs on `null`, it returns `null ` or `()`.

Reviewer: ```suggestion
    You can use optional field access to access values from a variable of type `json`. If the value is not a JSON object and is not null, optional field access returns an error. If the value is a JSON object that does not have a field with the specified key or if the value is null, it returns `null ` or `()`.
```

Author: Fixed.

    ```ballerina
    check person?.address?.city;
    ```

### JSON/Record/Map conversion

> Ballerina records, maps, and JSON objects hold records. Records are collections of fields, and you can access each field value by a key. You can write expressions converting from one type to another.

Reviewer: ```suggestion
Records and maps (including `map<json>` which represents JSON objects) are mapping that contain collections of fields. A field can be accessed using a key. You can write expressions converting from one type to another.
```

Author: Fixed.

###### Examples:

To convert the record `movieTheRevenant` to `json`, use the following:

```ballerina

>       check movieTheRevenant.cloneWithType(json);

Reviewer: I think we need to consider explaining `toJson` and `fromJsonWithType` too (or instead).

https://docs.central.ballerina.io/ballerina/lang.value/1.0.0/functions#toJson
https://docs.central.ballerina.io/ballerina/lang.value/1.0.0/functions#fromJsonWithType

Author: This is not yet supported by the Console. Hence took down the content.

      
```

> To convert the json `theRevenantJson` to record `Movie`, use the following:

Reviewer: ```suggestion
To convert the `theRevenantJson` variable of type `json` to the record `Movie`, use the following:
```

Author: Done.

# Ballerina Expression Syntax

Ballerina expressions comprise a powerful expression syntax based on the [Ballerina language](https://ballerina.io/) for method invocations, declaring a variable, etc. This document presents you with the common expressions that you can use while developing in the Choreo console.

## Variables
A variable declaration statement declares a value of a specific type (`string`, `integer`, etc.).  You can use a previously declared variable at a later phase in the logical process of your service or integration.

### int

> The `int` type is 64-bit signed integers (same as long in Java). A signed integer is an integer with the  `+` or `-` character in the front. However, You can also define a positive integer or zero using only numerical characters. The `int` type supports the usual [arithmetic operators: + - / %](#arithmetic-operators). The operator precedence is the same as C. Integer overflow is a runtime error in Ballerina. 

Reviewer 2: Why do we need to specifically say about java here?

Reviewer 3: In "However, You can also define.....", "you" should start with simple "y".

Author: @anupama-pathirage  This is the same description used in Ballerina docs.
@RukshiW  corrected.

> The `int` type is 64-bit signed integers (same as long in Java). A signed integer is an integer with the  `+` or `-` character in the front. However, You can also define a positive integer or zero using only numerical characters. The `int` type supports the usual [arithmetic operators: + - / %](#arithmetic-operators). The operator precedence is the same as C. Integer overflow is a runtime error in Ballerina. 

Reviewer 3: Since `int` is formatted with black ticks, the same should be done for "long".

Author: Done.

> The `int` type is 64-bit signed integers (same as long in Java). A signed integer is an integer with the  `+` or `-` character in the front. However, You can also define a positive integer or zero using only numerical characters. The `int` type supports the usual [arithmetic operators: + - / %](#arithmetic-operators). The operator precedence is the same as C. Integer overflow is a runtime error in Ballerina. 

Reviewer 3: Shall we change "The `in` type is 64-bit signed integers" to something like "The expressions of the `int` type are 64-bit integers..." so that the tense matches.

Author: Changed it to "The `int` type is a 64-bit two's complement integer (similar to `long` in Java)"

> ###### Examples:

Reviewer 3: Style guides generally recommend avoiding using punctuation marks in titles, and the external reviewer recommended the same.

Author: This is not used as a title, although we use it as a style. This does not appear as a title in any of the menus.

To initialize a variable of type `int` in Choreo, you can use one of the following expressions:

  ```ballerina
  1234
  
  -10
  
  +2020
  ```

### float

> The `float` type is IEEE 64-bit binary floating-point (same as double in Java) and supports the same arithmetic operators as `int`. You have the option to use `+` or `-` characters in the front to indicate the sign.

Reviewer 3: `double` should be formatted with black ticks.

Author: done.

###### Examples:

To initialize a variable of type `float` in Choreo, you can use one of the following expressions:

  ```ballerina
  0.3353
  
  -8593.992
  
  +950.930
  ```

### boolean

The boolean data type has one of the two possible values, `true` or `false`.

### string

> The `string` type represents an immutable sequence of zero or more Unicode characters. In Ballerina, there is no character type. Instead, a string of length is one represents a character. Two string values are considered equal (==) if both sequences have the same characters. You can use <, <=, >, and >= operators on string values. They work by comparing code points. Unpaired surrogates are not allowed.

Reviewer 3: "Instead, a string of length is one represents a character" is unclear and sounds grammatically incorrect.

Author: Fixed.

###### Examples:

To initialize a variable of type `string` in Choreo, you can use one of the following expressions:
  ```ballerina
  "Ballerina"
  
  "Ballerina is a programming language intended for network-distributed applications."
  ```

### var

> Ballerina allows you to declare variables using the `var` keyword. Here, Ballerina infers the type of the declared variable based on the right-hand side expression.

Reviewer 3: Suggestion: " Here, Ballerina infers the type of the declared variable based on the expression to the right."

Author: isn't "to" used for motion?

###### Examples:

To initialize a variable of type `var` in Choreo, you can use one of the following expressions:

```ballerina
5
    
"Hello!"
```

## Operators

### String operators

- Concatenation

    You can use the `+` operator to do string concatenations. The string value can come from a string literal or a variable.

    ###### Example:

    To concatenate the string `Hi` and a string variable `name`, you can use the following expression:

    ```ballerina
     "Hi " + name;
    ```

### Arithmetic operators

Arithmetic operators perform addition, subtraction, multiplication, division, and modulo on numerical values. You can use literals or variables of type `int` or `float` as the numerical values in Ballerina expressions.

- Addition

    You can use the `+` operator for number addition. 

    ###### Example:

    To add an integer literal (`20`) and an integer variable `number1`, you can use the following expression:

    ```ballerina
    number1 + 20;
    ```

- Subtraction 

    You can use the `-` operator for number subtraction. 

    ###### Example:

>     To subtract two integers, integer variable `yearOfBirth` from integer literal `2022`, you can use the following expression:

Reviewer 3: Suggestion: "To subtract two integers (e.g., integer variable `yearOfBirth` from integer literal `2022`),  you can use the following expression:"

    ```ballerina
    2022 - yearOfBirth;
    ```
 
- Division

    You can use the `/` operator for number division. 

    ###### Example:

    To divide an integer variable `totalNumberOfStudents` from integer literal `5`, you can use the following expression:

    ```ballerina
    totalNumberOfStudents / 5; 
    ```
 
- Multiplication

    You can use the `*` operator for number multiplication. 

    ###### Example:

    To multiply two float constants `60` by `5`, you can use the following expression:

    ```ballerina
    60 * 5;
    ```

- Modulo 

    You can use the  `%` operator to get the remainder of a division. 

    ###### Example:

    To get the remainder of the division between an integer variable `n` and an integer literal `3`, you can use the following expression:
  
    ```ballerina
    (n % 3) > 2
    ```

### Value/type operators

- Equality 

>     You can use the equality operators to check the equality of [Ballerina's](https://ballerina.io/) basic type literals. These values can come from a literal or a variable. **Note** that the resulting value from these operators is always of `boolean` type.   

Reviewer 3: Suggestion: "These values can be derived from a literal or a variable. The resulting value of these operators is always of the `boolean` type."

Reviewer 3: If the part about the resulting value needs to be highlighted as a note, can we use "!!! note"? Having notes in two forms can be inconsistent.

Author: fixed.

    
    Ballerina provides the following equality operators: 
    
    - `==`: For deep value equality of `anydata` typed values.
    - `===`: For reference equality. 
    - `!= `and `!==`: For negations.

    ###### Examples:

>     To compare an integer literal `10` with an integer variable `firstNumber`, you can use the following expression:

Reviewer 3: Suggestion: "To compare an integer literal (e.g., `10`) with an integer variable (e.g., `firstNumber`), you can use the following expression.

    
    ```ballerina
    10 == firstNumber
    ```

>     To compare a string variable `username`  with  a string literal `admin`, use the following expression:  

Reviewer 3: Suggestion: "To compare a string variable (e.g., `username`) with a string literal (e.g., `admin`), use the following expression."

    ```ballerina
    username != "admin" 
    ```

- Value comparison

    The following operators are available in the expression editor for value comparisons:
    
    - `<`: Less than
    - `>`: Greater than
    - `<=`: Less than or equal to
    - `>=`: Greater than or equal to

    ###### Examples:

    To compare the value of the integer variable `firstNumber` to the integer literal `10`, use one of the following expressions:

    ```ballerina
    firstNumber > 10
  
    firstNumber <= 10
    ```

    To compare the value of the string variable `greeting` to the string literal `Hi`, use the following expression:
   
    ```ballerina
    greeting == "Hi"    
    ```

>     To compare an array variable `arrayWithTwoElements` with an array literal `arrayWithOneElement`, use the following expression:

Reviewer: `arrayWithOneElement` is also a variable here?

Author: Yes. Corrected.

    
     ```ballerina
     arrayWithTwoElements <  arrayWithOneElement
     ```
    
## HTTP 

All the HTTP-related services created in Choreo have a variable called `request` (of type `http:Request`) which developers can use to query information related to the HTTP request. Following is a list of such use cases.

### Read a header value in the request

> A typical HTTP request contains many headers.  These include standard and custom headers. You can use the `request` variable's `getHeader` method to read a specific header value in any expression-enabled input field.

Reviewer 3: Remove extra space.

Author: Fixed

###### Examples:

To read the value of an HTTP header `Access-Control-Allow-Origin`, use the following expression:

```ballerina
check request.getHeader("Access-Control-Allow-Origin")
```
  
To read the value of an HTTP header by passing a variable that contains the header name `headerName`, use the following expression:

```ballerina
check request.getHeader(headerName)
```

### Check whether a header is present in the request

Services or integrations in Choreo sometimes need to check the availability of a header before proceeding to something like reading the header value. You can use the  `hasHeader`  method in the `request` object for this purpose.

###### Examples:

> To check if the HTTP header `content-length` exists given the header name, use the following expression:

Reviewer 3: Suggestion: "To check if the HTTP header `content-length` exists in the given header name...."

Author: Done

```ballerina
check request.getTextPayload()
```

### Read the value of a query parameter 

Query parameters are used in an HTTP request to send additional inputs to consider when processing the request. You can use the `request` variable's `getQueryParamValue` method to read the passed query parameter.

###### Examples:

To read the value of the query parameter `category`, use the following expression:

```ballerina
request.getQueryParamValue("category")
```
  
To read the value of the query parameter passing the query parameter name as a variable, use the following expression:

```ballerina
request.getQueryParamValue(queryParamName)
```

### Read the cookies available in the request

HTTP cookies keep stateful information against a client. You can use the `request` variable's `getCookies` method to read cookies present in the request as an array.

###### Example:

To read all the cookies available in the request, use the following expression:

```ballerina
request.getCookies()
```

## JSON

JSON is a textual format for representing a collection of values: a simple value, an array of values, or an object. 

### Simple values in JSON 

> Ballerina has a single type named `json` that can represent any JSON value. You can use `json` which is a built-in union type in Ballerina, to store values of the types `nil` (as the null value), `boolean`, `int`, `float`, `decimal`, `string`, `json[]`, and `map<json>`. 

Reviewer 3: The comma after "Ballerina" ios not needed.

Author: Fixed

```ballerina
[1, false, null, "foo", {first: "Alex", last: "John"}] 
```
To cast the `json` type variable `personRecord` to a `json[]` type variable, use the following:

```ballerina
<json[]>personRecord;
```

To access elements the fourth element in the JSON array `personRecordArray`, use one of the following:

!!! note
    If you define the JSON array using a `json` type variable, you need to cast it to a `json[]` type first.
    
```ballerina
personRecordArray[4]; 
```

To add a JSON array in an object literal, use the following:

```ballerina
    {
        fname: "Alex",
        lname: "Stallone",
        family: [
            {fname: "James", lname: "Stallone"},
            {fname: "Blake", lname: "Stallone"},
            {fname: "Kyle", lname: "Stallone"},
            {fname: "Taylor", lname: "Stallone"}
        ]
    }
```

To get the length of a JSON array, use the following:

```ballerina
personRecordArray.length(); 
```

### JSON access

You can use the field access(.) and optional field access (?.) to access the `json` type in Ballerina. In this case, the type checking happens at runtime. The return type of this operation is a union of `json` and `error`. 
    

> !!! info

Reviewer 3: Let's not use future tense here? 

Suggestion: " The examples in this section assume that the following variable of type `json` is already defined."

Author: Fixed.

Ballerina expressions comprise a powerful expression syntax based on the [Ballerina language](https://ballerina.io/) for function invocations, declaring a variable using literals and tuples, typecasting, etc. This document presents you with the common expressions that you can use while developing Choreo applications.

## Variables
A variable statement holds a value of a specific data type (`string`, `integer`, etc.).  You can declare a variable and use it later in the logical process of your service or the integration.

### var

Ballerina allows you to declare variables using the `var` keyword instead of specifying a type. The expression on the right-hand side decides the type of the declared variable.

###### Examples:

```ballerina

>     var varOfTypeInt = 5;

Reviewer 2: Remove indentation.

Author: done.

### JSON access

You can use the field access(.) and optional field access (?.) to access the `json` type in Ballerina. In this case, the type checking happens at runtime. The return type of this operation is a union of `json` and `error`. 
    
!!! info
    The examples in this section will assume the following json is already defined.
    
    ```ballerina
         json person = {
             fname: "Alex",
             lname: "Jane",
             address: {
                 line: "20 Palm Grove",
                 city: "Colombo 03",
                 country: "Sri Lanka"
             }
         }
    ```    

- Field access(.) 

    You can use the field access operator to access record fields of the `json` type. If the key is invalid or the field access occurs on `null`, it returns `error`.
   
    ###### Examples:

    To access a field, use the following:    

    ```ballerina
       json|error firstName = person.fname;
    ```
    To access a field in a nested level, use the following:
 
     ```ballerina
        json|error city = person.address.city;
     ```

- Optional field access (?.) 

>     You can use the optional field access operator to access record fields, including any optional fields. If the key is invalid or the field access occurs on `null`, it returns `null ` or `()`.

Reviewer 2: ```suggestion
    You can use the optional field access operator to access record fields, including any optional fields. If the key is invalid or the field access occurs on `null`, it returns `null` or `()`.
```

Author: done

    ```ballerina
    json|error year = j1?.age?.year;
    ```

For more information, refer to [Ballerina JSON Access](https://ballerina.io/learn/by-example/json-access.html)
 
### JSON to XML

You can convert JSON to using the `xmldata:fromJson()` function. This function takes a `xmldata:JsonOptions` record as an argument. The `JsonOptions` record has two fields: the `attributePrefix` and the `arrayEntryTag`. 
 - `attributePrefix`: A prefix that you use to identify the XML attributes. The default value of the `attributePrefix` option is `@`. 
 - `arrayEntryTag`: The name of the tag you add to each entry of the JSON array. The default value of the arrayEntryTag option is `root`.

!!! info
    The examples in this section will assume the following json is already defined.
    
    ```ballerina
    json storeDetails = {
         "Store": {
             "@id": "AST",
             "name": "Alex
             "address": {
                 "street": "Main",
                 "city": "94"
             },
             "codes": ["4", "8"]
         }
    };

>     ``` 

Reviewer 2: Will code snippets display properly within info. Github  view is not  doing that.

Author: ![Screenshot from 2021-06-04 11-48-03](https://user-images.githubusercontent.com/7173347/120754800-c63b3780-c52a-11eb-8a7b-18d79c253d6d.png)
The compiled view seems fine.
